### ISSUE TITLE
`manual_c_str_literals` suggestion in wrong place

### ISSUE THREAD

**Issue by YouKnow-sys** [rust-lang/rust-clippy#15936]
### Summary

clippy gave me this warning
```
  --> src/proxy.rs:39:19
   |
39 |             PCSTR(b"SomeApi\0".as_ptr()),
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^ help: use a `c""` literal: `c"SomeApi"`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_c_str_literals
   = note: `#[warn(clippy::manual_c_str_literals)]` on by default
```
when using `windows::core::PCSTR`, the thing is `PCSTR` expect the input to be `*const u8` but if we do what clippy is suggesting the `CStr::as_ptr` will give us `*const i8`... we can just use `cast` to cast the pointer to `*const u8` but overall this warning isn't correct in this situation.

### Lint Name

manual_c_str_literals

### Reproducer

I tried this code:

```rust
PCSTR(b"SomeApi\0".as_ptr())
```

I saw this happen:

```
  --> src/proxy.rs:39:19
   |
39 |             PCSTR(b"SomeApi\0".as_ptr()),
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^ help: use a `c""` literal: `c"SomeApi"`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_c_str_literals
   = note: `#[warn(clippy::manual_c_str_literals)]` on by default
```

I expected to see no clippy warning

### Version

```text
rustc 1.90.0 (1159e78c4 2025-09-14)
binary: rustc
commit-hash: 1159e78c4747b02ef996e55082b704c09b970588
commit-date: 2025-09-14
host: x86_64-unknown-linux-gnu
release: 1.90.0
LLVM version: 20.1.8
```

**Comment by ada4a** [comment]
This isn't really a false positive -- the suggestion _is_ correct, it's just missing the `.cast`. Looking at the lint source, the developers even acknowledge that a cast might be necessary, and for that reason make the suggestion non-auto-applicable: https://github.com/rust-lang/rust-clippy/blob/00f68d530d2131e8b3bb72c094c6ca18b825c4c1/clippy_lints/src/methods/manual_c_str_literals.rs#L44-L46

I guess the diagnostic could be improved a bit by basically emitting that comment using `Diag::note`

**Comment by YouKnow-sys** [comment]
> This isn't really a false positive -- the suggestion _is_ correct, it's just missing the `.cast`. Looking at the lint source, the developers even acknowledge that a cast might be necessary, and for that reason make the suggestion non-auto-applicable:
> 
> [rust-clippy/clippy_lints/src/methods/manual_c_str_literals.rs](https://github.com/rust-lang/rust-clippy/blob/00f68d530d2131e8b3bb72c094c6ca18b825c4c1/clippy_lints/src/methods/manual_c_str_literals.rs#L44-L46)
> 
> Lines 44 to 46 in [00f68d5](/rust-lang/rust-clippy/commit/00f68d530d2131e8b3bb72c094c6ca18b825c4c1)
> 
>  // an additional cast may be needed, since the type of `CStr::as_ptr` and 
>  // `"".as_ptr()` can differ and is platform dependent 
>  Applicability::HasPlaceholders, 
> I guess the diagnostic could be improved a bit by basically emitting that comment using `Diag::note`

I mean in my case applying the change will make the code invalid, so we can say it's not correct to suggest it in first place...

**Comment by ada4a** [comment]
> I mean in my case applying the change will make the code invalid, so we can say it's not correct to suggest it in first place...

No no, that's kind of expected for suggestion marked with `Applicability::MaybeIncorrect`[^1] -- they might bring you halfway there, but some additional fix-up might be required. Of course it would be better if Clippy would've suggested `PCSTR(c"SomeApi".as_ptr().cast())` directly, but I'd argue `PCSTR(c"SomeApi".as_ptr())` is already a step in the right direction -- the lint suggests the improvement it was supposed to be suggesting: avoiding manual nul-terminated strings.

[^1]: no idea why the developers used `Applicability::HasPlaceholders` here

**Comment by samueltardieu** [comment]
> the suggestion _is_ correct, it's just missing the `.cast`

Which makes it an incorrect suggestion IMO. This would be different if a note added that an additional cast is necessary (on selected platforms).
